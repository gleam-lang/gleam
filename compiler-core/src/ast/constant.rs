use super::*;
use crate::type_::{FieldMap, HasType};

pub type TypedConstant = Constant<Arc<Type>, SmolStr>;
pub type UntypedConstant = Constant<(), ()>;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Constant<T, RecordTag> {
    Int {
        location: SrcSpan,
        value: SmolStr,
    },

    Float {
        location: SrcSpan,
        value: SmolStr,
    },

    String {
        location: SrcSpan,
        value: SmolStr,
    },

    Tuple {
        location: SrcSpan,
        elements: Vec<Self>,
    },

    List {
        location: SrcSpan,
        elements: Vec<Self>,
        typ: T,
    },

    Record {
        location: SrcSpan,
        module: Option<SmolStr>,
        name: SmolStr,
        args: Vec<CallArg<Self>>,
        tag: RecordTag,
        typ: T,
        field_map: Option<FieldMap>,
    },

    BitString {
        location: SrcSpan,
        segments: Vec<BitStringSegment<Self, T>>,
    },

    Var {
        location: SrcSpan,
        module: Option<SmolStr>,
        name: SmolStr,
        constructor: Option<Box<ValueConstructor>>,
        typ: T,
    },
}

impl TypedConstant {
    pub fn type_(&self) -> Arc<Type> {
        match self {
            Constant::Int { .. } => crate::type_::int(),
            Constant::Float { .. } => crate::type_::float(),
            Constant::String { .. } => crate::type_::string(),
            Constant::BitString { .. } => crate::type_::bit_string(),
            Constant::Tuple { elements, .. } => {
                crate::type_::tuple(elements.iter().map(|e| e.type_()).collect())
            }
            Constant::List { typ, .. }
            | Constant::Record { typ, .. }
            | Constant::Var { typ, .. } => typ.clone(),
        }
    }
}

impl HasType for TypedConstant {
    fn type_(&self) -> Arc<Type> {
        self.type_()
    }
}

impl<A, B> Constant<A, B> {
    pub fn location(&self) -> SrcSpan {
        match self {
            Constant::Int { location, .. }
            | Constant::List { location, .. }
            | Constant::Float { location, .. }
            | Constant::Tuple { location, .. }
            | Constant::String { location, .. }
            | Constant::Record { location, .. }
            | Constant::BitString { location, .. }
            | Constant::Var { location, .. } => *location,
        }
    }

    pub fn is_simple(&self) -> bool {
        matches!(
            self,
            Self::Int { .. } | Self::Float { .. } | Self::String { .. }
        )
    }
}

impl<A, B> HasLocation for Constant<A, B> {
    fn location(&self) -> SrcSpan {
        self.location()
    }
}

impl<A, B> crate::bit_string::GetLitValue for Constant<A, B> {
    fn as_int_literal(&self) -> Option<i64> {
        if let Constant::Int { value, .. } = self {
            if let Ok(val) = value.parse::<i64>() {
                return Some(val);
            }
        }
        None
    }
}
