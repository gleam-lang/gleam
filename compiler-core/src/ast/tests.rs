use std::sync::Arc;

use crate::{
    ast::{SrcSpan, TypedExpr},
    type_::{
        self, AccessorsMap, Environment, ExprTyper, FieldMap, ModuleValueConstructor,
        RecordAccessor, Type, ValueConstructor, ValueConstructorVariant,
    },
    uid::UniqueIdGenerator,
};

use super::TypedModule;

fn compile_module(src: &str) -> TypedModule {
    use crate::type_::build_prelude;
    let (ast, _) = crate::parse::parse_module(src).expect("syntax error");
    let ids = UniqueIdGenerator::new();
    let mut modules = im::HashMap::new();
    // DUPE: preludeinsertion
    // TODO: Currently we do this here and also in the tests. It would be better
    // to have one place where we create all this required state for use in each
    // place.
    let _ = modules.insert("gleam".into(), build_prelude(&ids));
    crate::analyse::infer_module(
        crate::build::Target::Erlang,
        &ids,
        ast,
        crate::build::Origin::Src,
        &"thepackage".into(),
        &modules,
        &mut vec![],
    )
    .expect("should successfully infer")
}

fn compile_expression(src: &str) -> TypedExpr {
    let ast = crate::parse::parse_expression_sequence(src).expect("syntax error");

    let mut modules = im::HashMap::new();
    let ids = UniqueIdGenerator::new();
    // DUPE: preludeinsertion
    // TODO: Currently we do this here and also in the tests. It would be better
    // to have one place where we create all this required state for use in each
    // place.
    let _ = modules.insert("gleam".into(), type_::build_prelude(&ids));
    let mut warnings = vec![];
    let mut environment = Environment::new(ids, "mymod", &modules, &mut warnings);

    // Insert a cat record to use in the tests
    let cat_type = Arc::new(Type::App {
        public: true,
        module: "mymod".into(),
        name: "Cat".into(),
        args: vec![],
    });
    let variant = ValueConstructorVariant::Record {
        constructors_count: 1,
        name: "Cat".into(),
        arity: 2,
        location: SrcSpan { start: 12, end: 15 },
        field_map: Some(FieldMap {
            arity: 2,
            fields: [("name".into(), 0), ("age".into(), 1)].into(),
        }),
        module: "mymod".into(),
    };
    environment.insert_variable(
        "Cat".into(),
        variant,
        type_::fn_(vec![type_::string(), type_::int()], cat_type.clone()),
        true,
    );

    environment.insert_accessors(
        "Cat".into(),
        AccessorsMap {
            public: true,
            type_: cat_type,
            accessors: [
                (
                    "name".into(),
                    RecordAccessor {
                        index: 0,
                        label: "name".into(),
                        type_: type_::string(),
                    },
                ),
                (
                    "age".into(),
                    RecordAccessor {
                        index: 1,
                        label: "age".into(),
                        type_: type_::int(),
                    },
                ),
            ]
            .into(),
        },
    );
    ExprTyper::new(&mut environment)
        .infer(ast)
        .expect("should successfully infer")
}

#[test]
fn find_node_todo() {
    let expr = compile_expression(r#" todo "#);
    assert_eq!(expr.find_node(0), None);
    assert_eq!(expr.find_node(1), Some(&expr));
    assert_eq!(expr.find_node(4), Some(&expr));
    assert_eq!(expr.find_node(5), None);
}

#[test]
fn find_node_todo_with_string() {
    let expr = compile_expression(r#" todo("ok") "#);
    assert_eq!(expr.find_node(0), None);
    assert_eq!(expr.find_node(1), Some(&expr));
    assert_eq!(expr.find_node(10), Some(&expr));
    assert_eq!(expr.find_node(11), None);
}

#[test]
fn find_node_string() {
    let expr = compile_expression(r#" "ok" "#);
    assert_eq!(expr.find_node(0), None);
    assert_eq!(expr.find_node(1), Some(&expr));
    assert_eq!(expr.find_node(4), Some(&expr));
    assert_eq!(expr.find_node(5), None);
}

#[test]
fn find_node_float() {
    let expr = compile_expression(r#" 1.02 "#);
    assert_eq!(expr.find_node(0), None);
    assert_eq!(expr.find_node(1), Some(&expr));
    assert_eq!(expr.find_node(4), Some(&expr));
    assert_eq!(expr.find_node(5), None);
}

#[test]
fn find_node_int() {
    let expr = compile_expression(r#" 1302 "#);
    assert_eq!(expr.find_node(0), None);
    assert_eq!(expr.find_node(1), Some(&expr));
    assert_eq!(expr.find_node(4), Some(&expr));
    assert_eq!(expr.find_node(5), None);
}

#[test]
fn find_node_var() {
    let expr = compile_expression(
        r#"let wibble = 1
wibble"#,
    );

    let var = TypedExpr::Var {
        location: SrcSpan { start: 15, end: 21 },
        constructor: ValueConstructor {
            public: false,
            variant: ValueConstructorVariant::LocalVariable {
                location: SrcSpan { start: 4, end: 10 },
            },
            type_: type_::int(),
        },
        name: "wibble".into(),
    };

    assert_eq!(expr.find_node(14), None);
    assert_eq!(expr.find_node(15), Some(&var));
    assert_eq!(expr.find_node(20), Some(&var));
    assert_eq!(expr.find_node(21), None);
}

#[test]
fn find_node_sequence() {
    let sequence = compile_expression(r#"1 2 3"#);
    assert!(sequence.find_node(0).is_some());
    assert!(sequence.find_node(1).is_none());
    assert!(sequence.find_node(2).is_some());
    assert!(sequence.find_node(3).is_none());
    assert!(sequence.find_node(4).is_some());
    assert!(sequence.find_node(5).is_none());
}

#[test]
fn find_node_list() {
    let list = compile_expression(r#"[1, 2, 3]"#);

    let int1 = TypedExpr::Int {
        location: SrcSpan { start: 1, end: 2 },
        typ: type_::int(),
        value: "1".into(),
    };
    let int2 = TypedExpr::Int {
        location: SrcSpan { start: 4, end: 5 },
        typ: type_::int(),
        value: "2".into(),
    };
    let int3 = TypedExpr::Int {
        location: SrcSpan { start: 7, end: 8 },
        typ: type_::int(),
        value: "3".into(),
    };

    assert_eq!(list.find_node(0), Some(&list));
    assert_eq!(list.find_node(1), Some(&int1));
    assert_eq!(list.find_node(2), Some(&list));
    assert_eq!(list.find_node(3), Some(&list));
    assert_eq!(list.find_node(4), Some(&int2));
    assert_eq!(list.find_node(5), Some(&list));
    assert_eq!(list.find_node(6), Some(&list));
    assert_eq!(list.find_node(7), Some(&int3));
    assert_eq!(list.find_node(8), Some(&list));
    assert_eq!(list.find_node(9), None);
}

#[test]
fn find_node_tuple() {
    let tuple = compile_expression(r#"#(1, 2, 3)"#);

    let int1 = TypedExpr::Int {
        location: SrcSpan { start: 2, end: 3 },
        typ: type_::int(),
        value: "1".into(),
    };
    let int2 = TypedExpr::Int {
        location: SrcSpan { start: 5, end: 6 },
        typ: type_::int(),
        value: "2".into(),
    };
    let int3 = TypedExpr::Int {
        location: SrcSpan { start: 8, end: 9 },
        typ: type_::int(),
        value: "3".into(),
    };

    assert_eq!(tuple.find_node(0), Some(&tuple));
    assert_eq!(tuple.find_node(1), Some(&tuple));
    assert_eq!(tuple.find_node(2), Some(&int1));
    assert_eq!(tuple.find_node(3), Some(&tuple));
    assert_eq!(tuple.find_node(4), Some(&tuple));
    assert_eq!(tuple.find_node(5), Some(&int2));
    assert_eq!(tuple.find_node(6), Some(&tuple));
    assert_eq!(tuple.find_node(7), Some(&tuple));
    assert_eq!(tuple.find_node(8), Some(&int3));
    assert_eq!(tuple.find_node(9), Some(&tuple));
    assert_eq!(tuple.find_node(10), None);
}

#[test]
fn find_node_binop() {
    let expr = compile_expression(r#"1 + 2"#);
    assert!(expr.find_node(0).is_some());
    assert!(expr.find_node(1).is_none());
    assert!(expr.find_node(2).is_none());
    assert!(expr.find_node(3).is_none());
    assert!(expr.find_node(4).is_some());
    assert!(expr.find_node(5).is_none());
}

#[test]
fn find_node_tuple_index() {
    let expr = compile_expression(r#"#(1).0"#);

    let int = TypedExpr::Int {
        location: SrcSpan { start: 2, end: 3 },
        value: "1".into(),
        typ: type_::int(),
    };

    assert_eq!(expr.find_node(2), Some(&int));
    assert_eq!(expr.find_node(4), Some(&expr));
    assert_eq!(expr.find_node(5), Some(&expr));
    assert_eq!(expr.find_node(6), None);
}

#[test]
fn find_node_module_select() {
    let expr = TypedExpr::ModuleSelect {
        location: SrcSpan { start: 1, end: 3 },
        typ: type_::int(),
        label: "label".into(),
        module_name: "name".into(),
        module_alias: "alias".into(),
        constructor: ModuleValueConstructor::Fn {
            module: "module".into(),
            name: "function".into(),
            location: SrcSpan { start: 1, end: 55 },
        },
    };

    assert_eq!(expr.find_node(0), None);
    assert_eq!(expr.find_node(1), Some(&expr));
    assert_eq!(expr.find_node(2), Some(&expr));
    assert_eq!(expr.find_node(3), None);
}

#[test]
fn find_node_fn() {
    let expr = compile_expression("fn() { 1 }");

    let int = TypedExpr::Int {
        location: SrcSpan { start: 7, end: 8 },
        value: "1".into(),
        typ: type_::int(),
    };

    assert_eq!(expr.find_node(0), Some(&expr));
    assert_eq!(expr.find_node(6), Some(&expr));
    assert_eq!(expr.find_node(7), Some(&int));
    assert_eq!(expr.find_node(8), Some(&expr));
    assert_eq!(expr.find_node(9), Some(&expr));
    assert_eq!(expr.find_node(10), None);
}

#[test]
fn find_node_call() {
    let expr = compile_expression("fn(_, _) { 1 }(1, 2)");

    let retrn = TypedExpr::Int {
        location: SrcSpan { start: 11, end: 12 },
        value: "1".into(),
        typ: type_::int(),
    };

    let arg1 = TypedExpr::Int {
        location: SrcSpan { start: 15, end: 16 },
        value: "1".into(),
        typ: type_::int(),
    };

    let arg2 = TypedExpr::Int {
        location: SrcSpan { start: 18, end: 19 },
        value: "2".into(),
        typ: type_::int(),
    };

    assert_eq!(expr.find_node(11), Some(&retrn));
    assert_eq!(expr.find_node(14), Some(&expr));
    assert_eq!(expr.find_node(15), Some(&arg1));
    assert_eq!(expr.find_node(16), Some(&expr));
    assert_eq!(expr.find_node(18), Some(&arg2));
    assert_eq!(expr.find_node(19), Some(&expr));
}

#[test]
fn find_node_record_access() {
    let access = compile_expression(r#"Cat("Nubi", 3).name"#);

    let string = TypedExpr::String {
        location: SrcSpan { start: 4, end: 10 },
        value: "Nubi".into(),
        typ: type_::string(),
    };

    let int = TypedExpr::Int {
        location: SrcSpan { start: 12, end: 13 },
        value: "3".into(),
        typ: type_::int(),
    };

    assert_eq!(access.find_node(4), Some(&string));
    assert_eq!(access.find_node(9), Some(&string));
    assert_eq!(access.find_node(12), Some(&int));
    assert_eq!(access.find_node(14), Some(&access));
    assert_eq!(access.find_node(18), Some(&access));
    assert_eq!(access.find_node(19), None);
}

#[test]
fn find_node_record_update() {
    let update = compile_expression(r#"Cat(..Cat("Nubi", 3), age: 4)"#);

    let int = TypedExpr::Int {
        location: SrcSpan { start: 27, end: 28 },
        value: "4".into(),
        typ: type_::int(),
    };

    assert_eq!(update.find_node(0), Some(&update));
    assert_eq!(update.find_node(3), Some(&update));
    assert_eq!(update.find_node(27), Some(&int));
    assert_eq!(update.find_node(28), Some(&update));
    assert_eq!(update.find_node(29), None);
}

#[test]
fn find_node_try() {
    let try_ = compile_expression(r#"try x = Ok(1) Ok(2)"#);

    let int1 = TypedExpr::Int {
        location: SrcSpan { start: 11, end: 12 },
        value: "1".into(),
        typ: type_::int(),
    };

    let int2 = TypedExpr::Int {
        location: SrcSpan { start: 17, end: 18 },
        value: "2".into(),
        typ: type_::int(),
    };

    assert_eq!(try_.find_node(0), Some(&try_));
    assert_eq!(try_.find_node(11), Some(&int1));
    assert_eq!(try_.find_node(17), Some(&int2));
    assert_eq!(try_.find_node(19), None);
}

#[test]
fn find_node_case() {
    let case = compile_expression(
        r#"
case 1, 2 {
  _, _ -> 3
}
"#,
    );

    let int1 = TypedExpr::Int {
        location: SrcSpan { start: 6, end: 7 },
        value: "1".into(),
        typ: type_::int(),
    };

    let int2 = TypedExpr::Int {
        location: SrcSpan { start: 9, end: 10 },
        value: "2".into(),
        typ: type_::int(),
    };

    let int3 = TypedExpr::Int {
        location: SrcSpan { start: 23, end: 24 },
        value: "3".into(),
        typ: type_::int(),
    };

    assert_eq!(case.find_node(1), Some(&case));
    assert_eq!(case.find_node(6), Some(&int1));
    assert_eq!(case.find_node(9), Some(&int2));
    assert_eq!(case.find_node(23), Some(&int3));
    assert_eq!(case.find_node(25), Some(&case));
    assert_eq!(case.find_node(26), None);
}

#[test]
fn find_node_bool() {
    let negate = compile_expression(r#"!True"#);

    let bool = TypedExpr::Var {
        location: SrcSpan { start: 1, end: 5 },
        constructor: ValueConstructor {
            public: true,
            variant: ValueConstructorVariant::Record {
                constructors_count: 2,
                name: "True".into(),
                arity: 0,
                field_map: None,
                location: SrcSpan { start: 0, end: 0 },
                module: "".into(),
            },
            type_: type_::bool(),
        },
        name: "True".into(),
    };

    assert_eq!(negate.find_node(0), Some(&negate));
    assert_eq!(negate.find_node(1), Some(&bool));
    assert_eq!(negate.find_node(2), Some(&bool));
    assert_eq!(negate.find_node(3), Some(&bool));
    assert_eq!(negate.find_node(4), Some(&bool));
    assert_eq!(negate.find_node(5), None);
}

#[test]
fn find_node_statement_fn() {
    let module = compile_module(
        r#"

pub fn main() {
  Nil
}

"#,
    );

    assert!(module.find_node(0).is_none());
    assert!(module.find_node(1).is_none());

    // The fn
    assert!(module.find_node(2).is_some());
    assert!(module.find_node(14).is_some());

    assert!(module.find_node(15).is_none());
    assert!(module.find_node(16).is_none());
    assert!(module.find_node(17).is_none());
}

#[test]
fn find_node_statement_import() {
    let module = compile_module(
        r#"
import gleam
"#,
    );

    assert!(module.find_node(0).is_none());
    assert!(module.find_node(7).is_none());

    // The import
    assert!(module.find_node(8).is_some());
    assert!(module.find_node(12).is_some());

    assert!(module.find_node(13).is_none());
    assert!(module.find_node(14).is_none());
}

#[test]
fn find_node_use() {
    let use_ = compile_expression(
        r#"
use x <- fn(f) { f(1) }
124
"#,
    );

    assert!(use_.find_node(0).is_none());
    assert!(use_.find_node(1).is_some()); // The use
    assert!(use_.find_node(23).is_some());

    assert!(use_.find_node(26).is_some()); // The int
}
