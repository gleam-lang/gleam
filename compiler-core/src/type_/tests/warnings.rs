use super::*;
use crate::ast::TodoKind;
use crate::{assert_no_warnings, assert_warning};

#[test]
fn unknown_label() {
    // https://github.com/gleam-lang/gleam/issues/1098
    // calling function with unused labelled argument should not emit warnings
    assert_no_warnings!(
        r#"fn greet(name name: String, title _title: String) { name }
           pub fn main() { greet(name: "Sam", title: "Mr") }"#,
    );
}

#[test]
fn todo_warning_test() {
    assert_warning!(
        "fn main() { 1 == todo }",
        Warning::Todo {
            kind: TodoKind::Keyword,
            location: SrcSpan { start: 17, end: 21 },
            typ: Arc::new(Type::Var {
                type_: Arc::new(RefCell::new(TypeVar::Link { type_: int() })),
            }),
        },
    );
}

// https://github.com/gleam-lang/gleam/issues/1669
#[test]
fn todo_warning_correct_location() {
    assert_warning!(
        "pub fn main() {
        todo
      }"
    );
}

#[test]
fn todo_with_known_type() {
    assert_warning!(
        "pub fn main() -> String {
  todo
}"
    );
}

#[test]
fn empty_func_warning_test() {
    assert_warning!(
        "pub fn main() { foo() }
        pub fn foo() { }"
    );
}

#[test]
fn warning_variable_never_used_test() {
    assert_warning!(
        "
pub fn foo() { Ok(5) }
pub fn main() { let five = foo() }"
    );
}

#[test]
fn warning_private_function_never_used() {
    assert_warning!("fn main() { 5 }");
}

#[test]
fn warning_many_at_same_time() {
    assert_warning!(
        "
fn main() { let five = 5 }"
    );
}

#[test]
fn result_discard_warning_test() {
    // Implicitly discarded Results emit warnings
    assert_warning!(
        "
fn foo() { Ok(5) }
fn main() { foo(); 5 }",
        Warning::ImplicitlyDiscardedResult {
            location: SrcSpan { start: 32, end: 37 }
        }
    );
}

#[test]
fn result_discard_warning_test2() {
    // Explicitly discarded Results do not emit warnings
    assert_no_warnings!(
        "
pub fn foo() { Ok(5) }
pub fn main() { let _ = foo(); 5 }",
    );
}

#[test]
fn unused_int() {
    assert_warning!(
        "fn main() { 1; 2 }",
        Warning::UnusedLiteral {
            location: SrcSpan { start: 12, end: 13 }
        }
    );
}

#[test]
fn unused_float() {
    assert_warning!(
        "fn main() { 1.0; 2 }",
        Warning::UnusedLiteral {
            location: SrcSpan { start: 12, end: 15 }
        }
    );
}

#[test]
fn unused_string() {
    assert_warning!(
        "
    fn main() { 
        \"1\"; 2 
    }",
        Warning::UnusedLiteral {
            location: SrcSpan { start: 26, end: 29 }
        }
    );
}

#[test]
fn unused_bit_string() {
    assert_warning!(
        "
    fn main() { 
        <<3>>; 2 
    }",
        Warning::UnusedLiteral {
            location: SrcSpan { start: 26, end: 31 }
        }
    );
}

#[test]
fn unused_tuple() {
    assert_warning!(
        "
    fn main() { 
        #(1.0, \"Hello world\"); 2
    }",
        Warning::UnusedLiteral {
            location: SrcSpan { start: 26, end: 47 }
        }
    );
}

#[test]
fn unused_list() {
    assert_warning!(
        "
    fn main() { 
        [1, 2, 3]; 2 
    }",
        Warning::UnusedLiteral {
            location: SrcSpan { start: 26, end: 35 }
        }
    );
}

#[test]
fn record_update_warnings_test() {
    // Some fields are given in a record update do not emit warnings
    assert_no_warnings!(
        "
        pub type Person {
            Person(name: String, age: Int)
        };
        pub fn update_person() {
            let past = Person(\"Quinn\", 27)
            let present = Person(..past, name: \"Santi\")
            present
        }",
    );
}

#[test]
fn record_update_warnings_test2() {
    // No fields are given in a record update emit warnings
    assert_warning!(
        "
        pub type Person {
            Person(name: String, age: Int)
        };
        pub fn update_person() {
            let past = Person(\"Quinn\", 27)
            let present = Person(..past)
            present
        }",
        Warning::NoFieldsRecordUpdate {
            location: SrcSpan {
                start: 183,
                end: 197
            }
        }
    );
}

#[test]
fn record_update_warnings_test3() {
    // All fields given in a record update emits warnings
    assert_warning!(
        "
        pub type Person {
            Person(name: String, age: Int)
        };
        pub fn update_person() {
            let past = Person(\"Quinn\", 27)
            let present = Person(..past, name: \"Quinn\", age: 28)
            present
        }",
        Warning::AllFieldsRecordUpdate {
            location: SrcSpan {
                start: 183,
                end: 221
            }
        }
    );
}

#[test]
fn unused_private_type_warnings_test() {
    // External type
    assert_warning!(
        "external type X",
        Warning::UnusedType {
            name: "X".to_string(),
            location: SrcSpan { start: 0, end: 15 },
            imported: false
        }
    );
}

#[test]
fn unused_private_type_warnings_test2() {
    assert_no_warnings!("pub external type Y");
}

#[test]
fn unused_private_type_warnings_test3() {
    // Type alias
    assert_warning!(
        "type X = Int",
        Warning::UnusedType {
            name: "X".to_string(),
            location: SrcSpan { start: 0, end: 12 },
            imported: false
        }
    );
}

#[test]
fn unused_private_type_warnings_test4() {
    assert_no_warnings!("pub type Y = Int");
}

#[test]
fn unused_private_type_warnings_test5() {
    assert_no_warnings!("type Y = Int pub fn run(x: Y) { x }");
}

#[test]
fn unused_private_type_warnings_test6() {
    // Custom type
    assert_warning!(
        "type X { X }",
        Warning::UnusedConstructor {
            name: "X".to_string(),
            location: SrcSpan { start: 9, end: 10 },
            imported: false
        }
    );
}

#[test]
fn unused_private_type_warnings_test7() {
    assert_no_warnings!("pub type X { X }");
}

#[test]
fn unused_private_type_warnings_test8() {
    assert_no_warnings!("type X { X } pub fn a() { let b = X case b { X -> 1 } }");
}

#[test]
fn unused_private_fn_warnings_test() {
    assert_warning!(
        "fn a() { 1 }",
        Warning::UnusedPrivateFunction {
            name: "a".to_string(),
            location: SrcSpan { start: 0, end: 6 },
        }
    );
}

#[test]
fn used_private_fn_warnings_test() {
    assert_no_warnings!("pub fn a() { 1 }");
}

#[test]
fn used_private_fn_warnings_test2() {
    assert_no_warnings!("fn a() { 1 } pub fn b() { a }");
}

#[test]
fn unused_private_const_warnings_test() {
    assert_warning!(
        "const a = 1",
        Warning::UnusedPrivateModuleConstant {
            name: "a".to_string(),
            location: SrcSpan { start: 6, end: 7 },
        }
    );
}

#[test]
fn used_private_const_warnings_test() {
    assert_no_warnings!("pub const a = 1");
}

#[test]
fn used_private_const_warnings_test2() {
    assert_no_warnings!("const a = 1 pub fn b() { a }");
}

#[test]
fn unused_variable_warnings_test() {
    // function argument
    assert_warning!(
        "pub fn a(b) { 1 }",
        Warning::UnusedVariable {
            name: "b".to_string(),
            location: SrcSpan { start: 9, end: 10 },
        }
    );
}

#[test]
fn used_variable_warnings_test() {
    assert_no_warnings!("pub fn a(b) { b }");
}

#[test]
fn unused_variable_warnings_test2() {
    // Simple let
    assert_warning!(
        "pub fn a() { let b = 1 5 }",
        Warning::UnusedVariable {
            name: "b".to_string(),
            location: SrcSpan { start: 17, end: 18 },
        }
    );
}

#[test]
fn used_variable_warnings_test2() {
    assert_no_warnings!("pub fn a() { let b = 1 b }");
}

#[test]
fn unused_variable_shadowing_test() {
    assert_warning!(
        "pub fn a() { let b = 1 let b = 2 b }",
        Warning::UnusedVariable {
            name: "b".to_string(),
            location: SrcSpan { start: 17, end: 18 },
        }
    );
}

#[test]
fn used_variable_shadowing_test() {
    assert_no_warnings!("pub fn a() { let b = 1 let b = b + 1 b }");
}

#[test]
fn unused_destructure() {
    // Destructure
    assert_warning!(
        "pub fn a(b) { case b { #(c, _) -> 5 } }",
        Warning::UnusedVariable {
            name: "c".to_string(),
            location: SrcSpan { start: 25, end: 26 },
        }
    );
}

#[test]
fn used_destructure() {
    assert_no_warnings!("pub fn a(b) { case b { #(c, _) -> c } }");
}

#[test]
fn unused_imported_module_warnings_test() {
    assert_warning!(
        (
            vec!["gleam".to_string(), "foo".to_string()],
            "pub fn bar() { 1 }"
        ),
        "import gleam/foo",
        Warning::UnusedImportedModule {
            name: "foo".to_string(),
            location: SrcSpan { start: 7, end: 16 },
        }
    );
}

#[test]
fn unused_imported_module_with_alias_warnings_test() {
    assert_warning!(
        (
            vec!["gleam".to_string(), "foo".to_string()],
            "pub fn bar() { 1 }"
        ),
        "import gleam/foo as bar",
        Warning::UnusedImportedModule {
            name: "bar".to_string(),
            location: SrcSpan { start: 7, end: 23 },
        }
    );
}

#[test]
fn unused_imported_module_no_warning_on_used_function_test() {
    assert_no_warnings!(
        (
            vec!["gleam".to_string(), "foo".to_string()],
            "pub fn bar() { 1 }"
        ),
        "import gleam/foo; pub fn baz() { foo.bar() }",
    );
}

#[test]
fn unused_imported_module_no_warning_on_used_type_test() {
    assert_no_warnings!(
        (
            vec!["gleam".to_string(), "foo".to_string()],
            "pub type Foo = Int"
        ),
        "import gleam/foo; pub fn baz(a: foo.Foo) { a }",
    );
}

#[test]
fn unused_imported_module_no_warning_on_used_unqualified_function_test() {
    assert_no_warnings!(
        (
            vec!["gleam".to_string(), "foo".to_string()],
            "pub fn bar() { 1 }"
        ),
        "import gleam/foo.{bar}; pub fn baz() { bar() }",
    );
}

#[test]
fn unused_imported_module_no_warning_on_used_unqualified_type_test() {
    assert_no_warnings!(
        (
            vec!["gleam".to_string(), "foo".to_string()],
            "pub type Foo = Int"
        ),
        "import gleam/foo.{Foo}; pub fn baz(a: Foo) { a }",
    );
}

#[test]
fn module_access_registers_import_usage() {
    assert_no_warnings!(
        (
            vec!["gleam".to_string(), "bibble".to_string()],
            "pub const bobble = 1"
        ),
        "import gleam/bibble pub fn main() { bibble.bobble }",
    );
}

// https://github.com/gleam-lang/gleam/issues/978
#[test]
fn bit_pattern_var_use() {
    assert_no_warnings!(
        "
pub fn main(x) {
  let <<name_size:8, name:binary-size(name_size)>> = x
  name
}",
    );
}

// https://github.com/gleam-lang/gleam/issues/989
#[test]
fn alternative_case_clause_pattern_variable_usage() {
    assert_no_warnings!(
        "
pub fn main(s) {
  case s {
    [a] | [a, _] -> a
    _ -> 0 
  }
}"
    );
}

// https://github.com/gleam-lang/gleam/issues/1742
#[test]
fn imported_function_referenced_in_constant() {
    assert_no_warnings!(
        (vec!["one".to_string()], "pub fn two() { 2 }"),
        "
import one

pub const make_two = one.two
"
    );
}

// https://github.com/gleam-lang/gleam/issues/1742
#[test]
fn imported_constructor_referenced_in_constant() {
    assert_no_warnings!(
        (vec!["one".to_string()], "pub type Two { Two(Int) }"),
        "
import one

pub const make_two = one.Two
"
    );
}
