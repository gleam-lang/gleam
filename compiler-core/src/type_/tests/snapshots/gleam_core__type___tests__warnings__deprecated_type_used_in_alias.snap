---
source: compiler-core/src/type_/tests/warnings.rs
expression: "\n@deprecated(\"Don't use this!\")\npub type Cat {\n    Cat(name: String)\n}\n\npub type Dog = Cat\n        "
---
warning: Deprecated type used
  ┌─ /src/warning/wrn.gleam:7:16
  │
7 │ pub type Dog = Cat
  │                ^^^ This type has been deprecated

It was deprecated with this message: Don't use this!
